error: call to '__shfl' is ambiguous

Error in #796 
```
[build] FAILED: catch/catch_tests/unit/kernel/CMakeFiles/KernelTest.dir/hipShflTests.cc.o 
[build] /space/pvelesko/chipStar/hip-tests-update/build/bin/hipcc  -I/space/pvelesko/chipStar/hip-tests-update/CHIP -I/space/pvelesko/chipStar/hip-tests-update/PUBLIC -I/usr/include/level_zero -I/space/pvelesko/chipStar/hip-tests-update/hip-tests/catch/external/Catch2 -I/space/pvelesko/chipStar/hip-tests-update/hip-tests/catch/./include -I/space/pvelesko/chipStar/hip-tests-update/hip-tests/catch/./kernels -I/space/pvelesko/chipStar/hip-tests-update/build/include -I/space/pvelesko/chipStar/hip-tests-update/hip-tests/catch/external/picojson -I/space/pvelesko/chipStar/hip-tests-update/build/catch -Wno-duplicate-decl-specifier -Wno-tautological-constant-compare  -Wno-c++20-extensions -Wno-unused-result -Wno-delete-abstract-non-virtual-dtor -Wno-deprecated-declarations -Wunused-command-line-argument --std=c++17 -g -fPIE -Wno-format-extra-args -mf16c -Wall -O1 -std=c++17 -MD -MT catch/catch_tests/unit/kernel/CMakeFiles/KernelTest.dir/hipShflTests.cc.o -MF catch/catch_tests/unit/kernel/CMakeFiles/KernelTest.dir/hipShflTests.cc.o.d -o catch/catch_tests/unit/kernel/CMakeFiles/KernelTest.dir/hipShflTests.cc.o -c /space/pvelesko/chipStar/hip-tests-update/hip-tests/catch/unit/kernel/hipShflTests.cc
[build] /space/pvelesko/chipStar/hip-tests-update/hip-tests/catch/unit/kernel/hipShflTests.cc:38:58: error: call to '__shfl' is ambiguous
[build]    38 |     for (int j = 0; j < width; j++) out[i * width + j] = __shfl(val, j * width + i);
[build]       |                                                          ^~~~~~
[build] /space/pvelesko/chipStar/hip-tests-update/hip-tests/catch/unit/kernel/hipShflTests.cc:125:22: note: in instantiation of function template specialization 'matrixTranspose<__half>' requested here
[build]   125 |   hipLaunchKernelGGL(matrixTranspose<T>, dim3(1), dim3(THREADS_PER_BLOCK_X * THREADS_PER_BLOCK_Y),
[build]       |                      ^
[build] /space/pvelesko/chipStar/hip-tests-update/hip-tests/catch/unit/kernel/hipShflTests.cc:173:36: note: in instantiation of function template specialization 'runTest<__half>' requested here
[build]   173 |   SECTION("run test for __half") { runTest<__half>(); }
[build]       |                                    ^
[build] /space/pvelesko/chipStar/hip-tests-update/include/hip/devicelib/sync_and_util.hh:130:1: note: candidate function
[build]   130 | DECL_SHFL_OPS(int);
[build]       | ^
[build] /space/pvelesko/chipStar/hip-tests-update/include/hip/devicelib/sync_and_util.hh:121:27: note: expanded from macro 'DECL_SHFL_OPS'
[build]   121 |   extern __device__ TYPE_ __shfl(TYPE_ var, int srcLane,                       \
[build]       |                           ^
[build] /space/pvelesko/chipStar/hip-tests-update/include/hip/devicelib/sync_and_util.hh:131:1: note: candidate function
[build]   131 | DECL_SHFL_OPS(unsigned int);
[build]       | ^
[build] /space/pvelesko/chipStar/hip-tests-update/include/hip/devicelib/sync_and_util.hh:121:27: note: expanded from macro 'DECL_SHFL_OPS'
[build]   121 |   extern __device__ TYPE_ __shfl(TYPE_ var, int srcLane,                       \
[build]       |                           ^
[build] /space/pvelesko/chipStar/hip-tests-update/include/hip/devicelib/sync_and_util.hh:132:1: note: candidate function
[build]   132 | DECL_SHFL_OPS(long);
[build]       | ^
[build] /space/pvelesko/chipStar/hip-tests-update/include/hip/devicelib/sync_and_util.hh:121:27: note: expanded from macro 'DECL_SHFL_OPS'
[build]   121 |   extern __device__ TYPE_ __shfl(TYPE_ var, int srcLane,                       \
[build]       |                           ^
[build] /space/pvelesko/chipStar/hip-tests-update/include/hip/devicelib/sync_and_util.hh:133:1: note: candidate function
[build]   133 | DECL_SHFL_OPS(unsigned long);
[build]       | ^
[build] /space/pvelesko/chipStar/hip-tests-update/include/hip/devicelib/sync_and_util.hh:121:27: note: expanded from macro 'DECL_SHFL_OPS'
[build]   121 |   extern __device__ TYPE_ __shfl(TYPE_ var, int srcLane,                       \
[build]       |                           ^
[build] /space/pvelesko/chipStar/hip-tests-update/include/hip/devicelib/sync_and_util.hh:134:1: note: candidate function
[build]   134 | DECL_SHFL_OPS(float);
[build]       | ^
[build] /space/pvelesko/chipStar/hip-tests-update/include/hip/devicelib/sync_and_util.hh:121:27: note: expanded from macro 'DECL_SHFL_OPS'
[build]   121 |   extern __device__ TYPE_ __shfl(TYPE_ var, int srcLane,                       \
[build]       |                           ^
[build] /space/pvelesko/chipStar/hip-tests-update/include/hip/devicelib/sync_and_util.hh:135:1: note: candidate function
[build]   135 | DECL_SHFL_OPS(double);
[build]       | ^
[build] /space/pvelesko/chipStar/hip-tests-update/include/hip/devicelib/sync_and_util.hh:121:27: note: expanded from macro 'DECL_SHFL_OPS'
[build]   121 |   extern __device__ TYPE_ __shfl(TYPE_ var, int srcLane,                       \
[build]       |                           ^
[build] /space/pvelesko/chipStar/hip-tests-update/include/hip/devicelib/sync_and_util.hh:149:20: note: candidate function
[build]   149 | DECL_LONGLONG_SHFL(__shfl, int);
[build]       |                    ^
[build] /space/pvelesko/chipStar/hip-tests-update/include/hip/devicelib/sync_and_util.hh:149:20: note: candidate function
[build] 1 error generated when compiling for .
[build] 
[build] failed to execute:/space/pvelesko/install/llvm/17.0/bin/clang++ -include /space/pvelesko/chipStar/hip-tests-update/include/hip/spirv_fixups.h -I//space/pvelesko/chipStar/hip-tests-update/build/include -I/space/pvelesko/chipStar/hip-tests-update/CHIP -I/space/pvelesko/chipStar/hip-tests-update/PUBLIC -I/usr/include/level_zero -I/space/pvelesko/chipStar/hip-tests-update/hip-tests/catch/external/Catch2 -I/space/pvelesko/chipStar/hip-tests-update/hip-tests/catch/./include -I/space/pvelesko/chipStar/hip-tests-update/hip-tests/catch/./kernels -I/space/pvelesko/chipStar/hip-tests-update/build/include -I/space/pvelesko/chipStar/hip-tests-update/hip-tests/catch/external/picojson -I/space/pvelesko/chipStar/hip-tests-update/build/catch -Wno-duplicate-decl-specifier -Wno-tautological-constant-compare -Wno-c++20-extensions -Wno-unused-result -Wno-delete-abstract-non-virtual-dtor -Wno-deprecated-declarations -Wunused-command-line-argument --std=c++17 -g -fPIE -Wno-format-extra-args -mf16c -Wall -O1 -std=c++17 -MD -c -x hip /space/pvelesko/chipStar/hip-tests-update/hip-tests/catch/unit/kernel/hipShflTests.cc -D__HIP_PLATFORM_SPIRV__= --offload=spirv64 -nohipwrapperinc --hip-path=/space/pvelesko/chipStar/hip-tests-update/build --target=x86_64-unknown-linux-gnu   -I/space/pvelesko/chipStar/hip-tests-update/include -I/space/pvelesko/chipStar/hip-tests-update/HIP/include -I/space/pvelesko/chipStar/hip-tests-update/build/include -o catch/catch_tests/unit/kernel/CMakeFiles/KernelTest.dir/hipShflTests.cc.o -MT catch/catch_tests/unit/kernel/CMakeFiles/KernelTest.dir/hipShflTests.cc.o -MF catch/catch_tests/unit/kernel/CMakeFiles/KernelTest.dir/hipShflTests.cc.o.d
```

--- a/include/hip/spirv_hip_fp16.h
+++ b/include/hip/spirv_hip_fp16.h
@@ -1747,6 +1747,66 @@ THE SOFTWARE.
             }
         } // Anonymous namespace.
 
+        // Warp shuffle overloads for __half / __half2 (CHIP-SPV/chipStar#797).
+        // sync_and_util.hh only declares integer/float/double __shfl* overloads,
+        // so a __half argument was ambiguous (it converts to several of them).
+        // Implement the half variants by shuffling the raw bits through the
+        // existing integer __shfl* overloads, which is bit-exact and therefore
+        // value-preserving.
+        #define __CHIP_DEF_HALF_SHFL(OP_, SEL_T_)                              \
+          __device__ inline __half OP_(__half var, SEL_T_ selector,           \
+                                       int width = CHIP_DEFAULT_WARP_SIZE) {  \
+            unsigned short bits;                                               \
+            __builtin_memcpy(&bits, &var, sizeof(bits));                      \
+            int shuffled = OP_(static_cast<int>(bits), selector, width);      \
+            unsigned short out = static_cast<unsigned short>(shuffled);       \
+            __half result;                                                    \
+            __builtin_memcpy(&result, &out, sizeof(out));                     \
+            return result;                                                    \
+          }                                                                   \
+          __device__ inline __half2 OP_(__half2 var, SEL_T_ selector,         \
+                                        int width = CHIP_DEFAULT_WARP_SIZE) { \
+            int bits;                                                         \
+            __builtin_memcpy(&bits, &var, sizeof(bits));                     \
+            int shuffled = OP_(bits, selector, width);                       \
+            __half2 result;                                                   \
+            __builtin_memcpy(&result, &shuffled, sizeof(shuffled));          \
+            return result;                                                    \
+          }
+        __CHIP_DEF_HALF_SHFL(__shfl, int)
+        __CHIP_DEF_HALF_SHFL(__shfl_xor, int)
+        __CHIP_DEF_HALF_SHFL(__shfl_up, unsigned int)
+        __CHIP_DEF_HALF_SHFL(__shfl_down, unsigned int)
+        #undef __CHIP_DEF_HALF_SHFL
+
+        #define __CHIP_DEF_HALF_SHFL_SYNC(OP_, SEL_T_)                         \
+          __device__ inline __half OP_(unsigned mask, __half var,             \
+                                       SEL_T_ selector,                       \
+                                       int width = CHIP_DEFAULT_WARP_SIZE) {  \
+            unsigned short bits;                                               \
+            __builtin_memcpy(&bits, &var, sizeof(bits));                      \
+            int shuffled = OP_(mask, static_cast<int>(bits), selector, width);\
+            unsigned short out = static_cast<unsigned short>(shuffled);       \
+            __half result;                                                    \
+            __builtin_memcpy(&result, &out, sizeof(out));                     \
+            return result;                                                    \
+          }                                                                   \
+          __device__ inline __half2 OP_(unsigned mask, __half2 var,           \
+                                        SEL_T_ selector,                      \
+                                        int width = CHIP_DEFAULT_WARP_SIZE) { \
+            int bits;                                                         \
+            __builtin_memcpy(&bits, &var, sizeof(bits));                     \
+            int shuffled = OP_(mask, bits, selector, width);                 \
+            __half2 result;                                                   \
+            __builtin_memcpy(&result, &shuffled, sizeof(shuffled));          \
+            return result;                                                    \
+          }
+        __CHIP_DEF_HALF_SHFL_SYNC(__shfl_sync, int)
+        __CHIP_DEF_HALF_SHFL_SYNC(__shfl_xor_sync, int)
+        __CHIP_DEF_HALF_SHFL_SYNC(__shfl_up_sync, unsigned int)
+        __CHIP_DEF_HALF_SHFL_SYNC(__shfl_down_sync, unsigned int)
+        #undef __CHIP_DEF_HALF_SHFL_SYNC
+
         #if !defined(HIP_NO_HALF)
             using half = __half;
             using half2 = __half2;

--- a/tests/compiler/CMakeLists.txt
+++ b/tests/compiler/CMakeLists.txt
@@ -97,6 +97,7 @@ add_hipcc_test(TestFloat128Macros.hip TEST_NAME TestFloat128gnupp14
   HIPCC_OPTIONS -std=gnu++14 -fsyntax-only)
 
 add_hipcc_test(TestDoubleShuffleOverloads.hip HIPCC_OPTIONS -fsyntax-only)
+add_hipcc_test(TestFix797ShflHalf.hip HIPCC_OPTIONS -fsyntax-only)
 add_hipcc_test(TestHipComplexInclude.hip HIPCC_OPTIONS)
 
 add_hipcc_test(TestHipccAcceptCcFiles.cc HIPCC_OPTIONS)

--- a/tests/compiler/TestFix797ShflHalf.hip
+++ b/tests/compiler/TestFix797ShflHalf.hip
@@ -1,0 +1,35 @@
+// Regression test for CHIP-SPV/chipStar#797.
+// __shfl(__half, ...) was ambiguous because only integer/float/double
+// overloads were declared; __half implicitly converts to several of them.
+#include <hip/hip_runtime.h>
+#include <hip/hip_fp16.h>
+
+__global__ void callShflHalf(__half *InOut, int LaneSel) {
+  *InOut = __shfl(*InOut, LaneSel);
+}
+__global__ void callShflUpHalf(__half *InOut, int LaneSel) {
+  *InOut = __shfl_up(*InOut, LaneSel);
+}
+__global__ void callShflDownHalf(__half *InOut, int LaneSel) {
+  *InOut = __shfl_down(*InOut, LaneSel);
+}
+__global__ void callShflXorHalf(__half *InOut, int LaneSel) {
+  *InOut = __shfl_xor(*InOut, LaneSel);
+}
+
+__global__ void callShflHalf2(__half2 *InOut, int LaneSel) {
+  *InOut = __shfl(*InOut, LaneSel);
+}
+
+__global__ void callShflSyncHalf(__half *InOut, int LaneSel) {
+  *InOut = __shfl_sync(0xffffffff, *InOut, LaneSel);
+}
+__global__ void callShflUpSyncHalf(__half *InOut, int LaneSel) {
+  *InOut = __shfl_up_sync(0xffffffff, *InOut, LaneSel);
+}
+__global__ void callShflDownSyncHalf(__half *InOut, int LaneSel) {
+  *InOut = __shfl_down_sync(0xffffffff, *InOut, LaneSel);
+}
+__global__ void callShflXorSyncHalf(__half *InOut, int LaneSel) {
+  *InOut = __shfl_xor_sync(0xffffffff, *InOut, LaneSel);
+}

